src/libServer/GetworkServer.cpp

```diff
@@ -0,0 +1,286 @@
+    return FAIL_RESULT;
+  }
+
+  return ethash_mining_result_t{POW::BlockhashToHexString(final_result),
+                                mixdigest, winning_nonce, true};
+}
+
+// UpdateCurrentResult check and update new result
+bool GetWorkServer::UpdateCurrentResult(
+    const ethash_mining_result_t& newResult) {
+  if (!newResult.success) {
+    LOG_GENERAL(WARNING, "newResult is not success");
+    return false;
+  }
+
+  bool accept = false;
+
+  lock_guard<mutex> g(m_mutexResult);
+
+  if (!m_curResult.success) {
+    // accept the new result directly if current result is false
+    accept = true;
+  } else {
+    // accept the new result if it less or equal than current one
+    auto new_hash = POW::StringToBlockhash(newResult.result);
+    auto cur_hash = POW::StringToBlockhash(m_curResult.result);
+    accept = ethash::is_less_or_equal(new_hash, cur_hash);
+  }
+
+  if (accept) {
+    // save new result, and notify other threads to get new result
+    m_curResult = newResult;
+    m_cvGotResult.notify_all();
+    LOG_GENERAL(INFO, "newResult accepted!");
+  } else {
+    LOG_GENERAL(INFO, "newResult is not accepted!");
+  }
+
+  return accept;
+}
+
+//////////////////////////////////////////////////
+// RPC Methods
+//////////////////////////////////////////////////
+
+// ETH getWork Server
+Json::Value GetWorkServer::getWork() {
+  LOG_MARKER();
+  Json::Value result;
+
+  lock_guard<mutex> g(m_mutexWork);
+
+  result.append(m_isMining ? m_curWork.header : "");
+  result.append(m_isMining ? m_curWork.seed : "");
+  result.append(m_isMining ? m_curWork.boundary : "");
+  result.append(m_isMining.load());
+  result.append(GetSecondsToNextPoW());
+
+  return result;
+}
+
+bool GetWorkServer::submitWork(const string& _nonce, const string& _header,
+                               const string& _mixdigest,
+                               const string& _boundary,
+                               const string& _miner_wallet,
+                               const string& _worker) {
+  (void)_miner_wallet;
+  (void)_worker;
+
+  LOG_MARKER();
+
+  if (!m_isMining) {
+    LOG_GENERAL(WARNING, "PoW is not running, ignore submit");
+    return false;
+  }
+
+  string nonce = _nonce;
+  string header = _header;
+  string mixdigest = _mixdigest;
+  string boundary = _boundary;
+
+  LOG_GENERAL(INFO, "Got PoW Result: ");
+  LOG_GENERAL(INFO, "    nonce: " << nonce);
+  LOG_GENERAL(INFO, "    header: " << header);
+  LOG_GENERAL(INFO, "    mixdigest: " << mixdigest);
+  LOG_GENERAL(INFO, "    boundary: " << boundary);
+
+  if (!DataConversion::NormalizeHexString(nonce) ||
+      !DataConversion::NormalizeHexString(header) ||
+      !DataConversion::NormalizeHexString(mixdigest) ||
+      !DataConversion::NormalizeHexString(boundary)) {
+    LOG_GENERAL(WARNING, "Invalid input parameters");
+    return false;
+  }
+
+  auto result = VerifySubmit(nonce, header, mixdigest, boundary);
+
+  bool accepted = UpdateCurrentResult(result);
```

Comment:
Sometimes our clang complains of unnecessary temp variables, I wonder why it didn't complain here.  But still you can directly return `UpdateCurrentResult` instead of going through `accepted`